tools/mkcomposefs.c

```diff
@@ -893,43 +877,318 @@ static ssize_t write_cb(void *_file, void *buf, size_t count)
-		if (ret < 0)
-			return ret;
+		if (join_paths(&tmp_path, path, fname) < 0)
+			return -1;
+
 		path = tmp_path;
 	}
 
 	if (lcfs_node_dirp(node)) {
-		size_t n_children = lcfs_node_get_n_children(node);
+		const size_t n_children = lcfs_node_get_n_children(node);
 		for (size_t i = 0; i < n_children; i++) {
-			struct lcfs_node_s *child = lcfs_node_get_child(node, i);
-			ret = fill_store(child, path, digest_store_path);
-			if (ret < 0)
-				return ret;
+			if (construct_copy_data(lcfs_node_get_child(node, i),
+						collection, path) < 0) {
+				return -1;
+			}
 		}
 	} else if ((lcfs_node_get_mode(node) & S_IFMT) == S_IFREG &&
 		   lcfs_node_get_content(node) == NULL &&
 		   lcfs_node_get_payload(node) != NULL) {
-		const char *payload = lcfs_node_get_payload(node);
+		if (add_to_work_collection(collection, node, path) < 0) {
+			return -1;
+		}
+	}
 
-		ret = copy_file_with_dirs_if_needed(path, digest_store_path,
-						    payload, true);
-		if (ret < 0)
-			return ret;
+	return 0;
+}
+
+static int construct_compute_data(struct lcfs_node_s *node,
+				  struct work_collection *collection,
+				  const char *path)
+{
+	cleanup_free char *tmp_path = NULL;
+	const char *fname = lcfs_node_get_name(node);
+
+	if (fname) {
+		if (join_paths(&tmp_path, path, fname) < 0)
+			return -1;
+
+		path = tmp_path;
+	}
+
+	if ((node->inode.st_mode & S_IFMT) == S_IFREG) {
+		if (add_to_work_collection(collection, node, path) < 0) {
+			return -1;
+		}
+	}
+
+	if (!lcfs_node_dirp(node))
+		return 0;
+
+	size_t n_children = lcfs_node_get_n_children(node);
+	for (size_t i = 0; i < n_children; i++) {
+		struct lcfs_node_s *child = lcfs_node_get_child(node, i);
+		if (construct_compute_data(child, collection, path) < 0) {
+			return -1;
+		}
 	}
 
 	return 0;
 }
 
+struct work_item_iterator {
+	pthread_mutex_t mutex_node_iterator;
+	int current_item;
+	int errorcode;
+	bool cancel_request;
+};
+
+static struct work_item *get_next_work_item(struct work_collection *collection,
+					    struct work_item_iterator *iterator)
+{
+	if (!iterator || !collection)
+		return NULL;
+
+	bool cancel = false;
+	struct work_item *ret = NULL;
+
+	pthread_mutex_lock(&(iterator->mutex_node_iterator));
+	if (iterator->cancel_request)
+		cancel = true;
+	else if (iterator->current_item < collection->count) {
+		ret = &(collection->items[iterator->current_item]);
+		iterator->current_item++;
+	}
+	pthread_mutex_unlock(&(iterator->mutex_node_iterator));
+	return cancel ? NULL : ret;
+}
+
+static void request_cancel(struct work_item_iterator *iterator, int errorcode)
+{
+	pthread_mutex_lock(&(iterator->mutex_node_iterator));
+	iterator->cancel_request = true;
+	iterator->errorcode = errorcode;
+	pthread_mutex_unlock(&(iterator->mutex_node_iterator));
+}
+
+typedef int (*THREAD_PROCESS_PROC)(struct work_item *, void *);
+
+static int process_copy(struct work_item *item, void *digest_store_path)
+{
+	if (copy_file_with_dirs_if_needed(item->path, (const char *)digest_store_path,
```

Comment:
I don't think you need the if here. You can just: "return copy_file_with_dirs_if_needed(...);"

Reply:
that was a miss. fixed